File: pypp_cli/test_dir/python/src/inconsistent_behviour/reassigning_a_ref.py
from pypp_python import Ref


# Rule of thumb:
# - do not reassign a reference variable.


# Implication:
# If you have a global state (say a list), and somewhere in your huge codebase
# you have a functions whos argument is a reference to that list, and you want to
# completely reassign it


def modifying_a_ref(l: list[int]):
    l = [-1, 3]


def reassigning_a_ref_fn():
    print("REASSIGNING A REF RESULTS:")
    a: list[int] = [1, 2, 3]
    b: Ref(list[int]) = a
    b.append(4)
    print(a)
    print(b)
    # Now, what if we reassign b to a new list?
    b = [10, 20, 30]  # This is not modifying the original list a!
    print(f"{a} (should be [10, 20, 30] in C++)")  # should print [1, 2, 3, 4]
    print(b)  # should print [10, 20, 30]

    # example 2
    c: list[int] = [1, 2, 3]
    modifying_a_ref(c)
    print(f"{c} (should be [-1, 3] in C++)")  # should print [1, 2, 3]


# Solution to this problem

File: pypp_cli/test_dir/cpp/src/inconsistent_behviour/reassigning_a_ref.cpp
#include "inconsistent_behviour/reassigning_a_ref.h"
#include "py_str.h"
#include "pypp_util/print.h"

namespace me {
void modifying_a_ref(pypp::PyList<int> &l) {
    l = pypp::PyList({-1, 3});
    pypp::print(l);
}

void reassigning_a_ref_fn() {
    pypp::print(pypp::PyStr("REASSIGNING A REF RESULTS:"));
    pypp::PyList<int> a({1, 2, 3});
    pypp::PyList<int> &b = a;
    b.append(4);
    pypp::print(a);
    pypp::print(b);
    b = pypp::PyList({10, 20, 30});
    pypp::print(
        pypp::PyStr(std::format("{} (should be [10, 20, 30] in C++)", a)));
    pypp::print(b);
    pypp::PyList<int> c({1, 2, 3});
    modifying_a_ref(c);
    pypp::print(pypp::PyStr(std::format("{} (should be [-1, 3] in C++)", c)));
}

} // namespace me